ASoC: Intel: sof_sdw: move controls and widgets addition to sof_sdw_rtd_init

bardliao opened the pull request:
Currently, we add card controls and dapm widgets one by one in the
codec_info->dais->rtd_init callback. Duplicated controls and dapm
widgets will be added if there are more than one types of amps in the
dai link. Moving it to sof_sdw_rtd_init() and only add the
controls/widgets of the first codec dai can avoid the duplications.
And add "spk:" tag to components string for each amp dai type.

bardliao committed:
ASoC: Intel: sof_sdw: add controls and dapm widgets in codec_info

Currently, we add card controls and dapm widgets one by one in the
codec_info->dais->rtd_init callback. Duplicated controls and dapm
widgets will be added if there are more than one types of amps in the
dai link. Moving it to sof_sdw_rtd_init() and only add the
controls/widgets of the first codec dai can avoid the duplications.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -1277,6 +1278,34 @@ static int sof_sdw_rtd_init(struct snd_soc_pcm_runtime *rtd)
+					return ret;
+				}
+			}
+		}
```
could we move all this to line 1272? This is a bit orthogonal to the rtd stuff and now it's in the middle. It'd be cleaner if the two parts didn't interfere.

bardliao replied:
> could we move all this to line 1272? This is a bit orthogonal to the rtd stuff and now it's in the middle. It'd be cleaner if the two parts didn't interfere.

No, we can't add the controls/widgets when rtd_init_done is true. Otherwise, it will be done twice for both playback and capture links. Those additions were done in `codec_info->dais[dai_index].rtd_init(rtd)`, we just move them outside the .rtd_init, but don't want to change the behavior.

ujfalusi replied:
Arguably you could also save the snd_soc_dapm_route for like with controls/widgets and do the snd_soc_dapm_add_routes() in this level.

bardliao replied:
> Arguably you could also save the snd_soc_dapm_route for like with controls/widgets and do the snd_soc_dapm_add_routes() in this level.

Right, but routes are very codec specific. So, I prefer keep it in the .rtd_init callback.

ujfalusi replied:
The widgets and controls also codec specific in some cases...

bardliao replied:
> The widgets and controls also codec specific in some cases...

The controls/widgets are at card level, and we can use our own name. But, routes need to map to the input/output widgets in codec drivers. Besides, we only want to add the controls/widgets once in a link, but we need to add the routes for each codec dai.

plbossart commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -711,10 +711,10 @@ static struct sof_sdw_codec_info codec_info_list[] = {
+				.controls = generic_jack_controls,
+				.num_controls = ARRAY_SIZE(generic_jack_controls),
+				.widgets = generic_jack_widgets,
+				.num_widgets = ARRAY_SIZE(generic_jack_widgets),
```
I would have removed redundant controls/widgets and use common ones first, then introduce the changes. It'd be easier to review/test.

plbossart commented on sound/soc/intel/boards/sof_sdw_rt_amp.c:
```diff
@@ -183,16 +183,31 @@ int rt_amp_spk_rtd_init(struct snd_soc_pcm_runtime *rtd)
+		if (!strstr(dai->name, "rt13"))
+			continue;
+		snprintf(codec_name, CODEC_NAME_SIZE, "%s", dai->name);
+		c_temp = devm_kasprintf(card->dev, GFP_KERNEL, "spk:%s", codec_name);
```
test allocation of c_temp

bardliao replied:
done

plbossart commented on sound/soc/intel/boards/sof_sdw_rt_amp.c:
```diff
@@ -183,16 +183,31 @@ int rt_amp_spk_rtd_init(struct snd_soc_pcm_runtime *rtd)
+		if (!strstr(dai->name, "rt13"))
+			continue;
+		snprintf(codec_name, CODEC_NAME_SIZE, "%s", dai->name);
+		c_temp = devm_kasprintf(card->dev, GFP_KERNEL, "spk:%s", codec_name);
```
also do we even need codec_name, could we simplify with

c_temp = devm_kasprintf(card->dev, GFP_KERNEL, "spk:%s", dai->name);

bardliao replied:
> also do we even need codec_name, could we simplify with
> 
> c_temp = devm_kasprintf(card->dev, GFP_KERNEL, "spk:%s", dai->name);

We will check if the tag is already in card->components below. So, if we use dai->name, we will add two spk: tags if different dais are used. It is not likely to happen. But, who knows.

plbossart replied:
I was only suggesting combining lines 200 and 201 in a single line, then you don't need codec_name as a variable

bardliao replied:
> I was only suggesting combining lines 200 and 201 in a single line, then you don't need codec_name as a variable

The purpose of `snprintf(codec_name, CODEC_NAME_SIZE, "%s", dai->name);` is to get the codec_name from the first 6 letters of dai->name. I can't find `devm_kasnprintf` to construct c_temp in one step.

charleskeepax commented on sound/soc/intel/boards/sof_sdw_common.h:
```diff
@@ -79,13 +79,65 @@ enum {
 
+static const struct snd_soc_dapm_widget cs_spk_widgets[] = {
+	SND_SOC_DAPM_SPK("Speakers", NULL),
+};
```
I don't really see any problem with updating the cirrus parts to use the generic speaker widget. The name of the widget does not currently form any sort of ABI and we will gain a PIN_SWITCH we probably should have anyway. Minor update would be needed in sof_sdw_cs_amp, but that is it. Unless @rfvirgil has any objections?

rfvirgil replied:
I agree.  "Speakers" -> "Speaker" doesn't matter, there isn't anything depending on the name.
So sof_sdw_cs_amp can use generic_spk_widgets.

bardliao replied:
> I agree. "Speakers" -> "Speaker" doesn't matter, there isn't anything depending on the name. So sof_sdw_cs_amp can use generic_spk_widgets.

Thanks for the comment. Let me change it to generic_spk_widgets.

charleskeepax commented on sound/soc/intel/boards/sof_sdw_common.h:
```diff
@@ -79,13 +79,65 @@ enum {
+
+static const struct snd_soc_dapm_widget cs42l43_dmic_widgets[] = {
+	SND_SOC_DAPM_MIC("DMIC", NULL),
+};
```
Would it make sense to just make a generic onboard mic option and update cs42l43 to use that? Probably with a PIN_SWITCH too. Also do we actually need to differentiate between AMIC and DMIC? Maybe we should just all it "On-board Mic" or something?

bardliao replied:
> Would it make sense to just make a generic onboard mic option and update cs42l43 to use that? Probably with a PIN_SWITCH too. Also do we actually need to differentiate between AMIC and DMIC? Maybe we should just all it "On-board Mic" or something?

renamed cs42l43_dmic_widgets to generic_dmic_widgets.

charleskeepax commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -1277,6 +1346,34 @@ static int sof_sdw_rtd_init(struct snd_soc_pcm_runtime *rtd)
+		 */
+		if (i == 0) {
+			const struct snd_soc_dapm_widget *w = codec_info->dais[dai_index].widgets;
+			const struct snd_kcontrol_new *c = codec_info->dais[dai_index].controls;
```
Minor nitpick but I am not a huge fan of single letter variable names outside of loop counters, but willing to look the other way if everyone else is happy.

Also a thought (more of discussion point than definitely saying we should do this), but would there be mileage in using the generic widgets/controls based on dai_type if the link doesn't define custom ones? That way you every new dai_link wouldn't need to add links to the generic ones. Really it would also be nice if all new code used the generic ones unless there was a very clear reason not to. Only slight difficulties I see are the realtek mic codecs that currently define no widgets although that seems weird since they presumably do have a mic associated and the mock up codecs. Although in all those cases the dangling widgets will be redundant but not cause any issues.

bardliao replied:
> Minor nitpick but I am not a huge fan of single letter variable names outside of loop counters, but willing to look the other way if everyone else is happy.

I use a single letter is for the 100-character limitation. I am happy if someone can suggest a better way to fit the limitation.

> 
> Also a thought (more of discussion point than definitely saying we should do this), but would there be mileage in using the generic widgets/controls based on dai_type if the link doesn't define custom ones? That way you every new dai_link wouldn't need to add links to the generic ones. Really it would also be nice if all new code used the generic ones unless there was a very clear reason not to. Only slight difficulties I see are the realtek mic codecs that currently define no widgets although that seems weird since they presumably do have a mic associated and the mock up codecs. Although in all those cases the dangling widgets will be redundant but not cause any issues.

Yes, I thought about it, too. However, I saw not all codecs eg. cs35l56 add their own controls. So, I don't add controls/widgets by default for now.

plbossart replied:
> I use a single letter is for the 100-character limitation. I am happy if someone can suggest a better way to fit the limitation.

usual gotos to avoid indentation...

if (i > 0)
    goto label;

do stuff

label:

bardliao replied:
updated.

bardliao committed:
ASoC: Intel: sof_sdw_cs_amp: rename Speakers to Speaker

To be consistent to other amps that used in the machine driver.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: sof_sdw: use generic name for controls/widgets

Some controls and widgets have exactly the same items. Rename them to be
generic. This is a preparation for further cleanup. No function change.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

ujfalusi commented on sound/soc/intel/boards/sof_sdw_common.h:
```diff
@@ -79,6 +79,24 @@ enum {
 
 #define SOF_SDW_MAX_DAI_NUM		8
 
+static const struct snd_soc_dapm_widget generic_spk_widgets[] = {
```
typo in commit message: s/wiegets/widgets

bardliao replied:
> typo in commit message: s/wiegets/widgets

fixed

ujfalusi commented on sound/soc/intel/boards/sof_sdw_common.h:
```diff
@@ -79,6 +79,24 @@ enum {
+static const struct snd_kcontrol_new generic_jack_controls[] = {
+	SOC_DAPM_PIN_SWITCH("Headphone"),
+	SOC_DAPM_PIN_SWITCH("Headset Mic"),
+};
```
So, these will be added to data each time the sof_sdw_common.h is included (sure, included and used, but I think 9 times with this patch)

bardliao replied:
> So, these will be added to data each time the sof_sdw_common.h is included (sure, included and used, but I think 9 times with this patch)

Maybe move the stuff to sof_sdw.c?

bardliao replied:
Let me move to sof_sdw.c to avoid errors like 
```
sound/soc/intel/boards/sof_sdw_common.h:121:38: error: ‘generic_jack_controls’ defined but not used [-Werror=unused-const-variable=]
  121 | static const struct snd_kcontrol_new generic_jack_controls[] = {
      | 
```

ujfalusi commented on sound/soc/intel/boards/sof_sdw_common.h:
```diff
@@ -79,6 +79,32 @@ enum {
 
 #define SOF_SDW_MAX_DAI_NUM		8
 
+static const struct snd_soc_dapm_widget cs42l43_dmic_widgets[] = {
```
typo in commit message: s/plave/place

bardliao replied:
> typo in commit message: s/plave/place

fixed

ujfalusi commented on sound/soc/intel/boards/sof_sdw_common.h:
```diff
@@ -79,6 +79,32 @@ enum {
+static const struct snd_soc_dapm_widget maxim_widgets[] = {
+	SND_SOC_DAPM_SPK("Left Spk", NULL),
+	SND_SOC_DAPM_SPK("Right Spk", NULL),
+};
```
what is the purpose of this move? If these are codec specific then they should be remain where they were? Would this again increase data size?

bardliao replied:
> what is the purpose of this move? If these are codec specific then they should be remain where they were? Would this again increase data size?

This commit is more like a preparation for next commit to add controls/widgets in sof_sdw_rtd_init()

ujfalusi commented on sound/soc/intel/boards/sof_sdw_rt_amp.c:
```diff
@@ -183,16 +183,34 @@ int rt_amp_spk_rtd_init(struct snd_soc_pcm_runtime *rtd)
+	for_each_rtd_codec_dais(rtd, i, dai) {
+		/*
+		 * rt_amp_spk_rtd_init() only for rt13xx amps. rt7xx multi-function codecs will
+		 * use their own rtd_init function.
```
this comment confused me a lot...
multi-function codecs, like rt7xx will not use the `rt_amp_spk_rtd_init` but their own rtd_init function since they need to add slightly different routes (I guess).
So this comment and the check for name might be redundant (and confusing)?

bardliao replied:
@ujfalusi "spk:rt7xx" will be added by `rt7xx_spk_rtd_init`, and we don't want to add it here. And here we add "spk:%s" for each codec dai, so we need to check the dai->name and only add "spk:rt13xx" to card->components.

ujfalusi replied:
But `spk:rt7xx` is added by `rt7xx_spk_rtd_init` (where xx is 12 or 22 atm, right?) and in those case the `rt_amp_spk_rtd_init` will not be called, or am I missing something?
For me it looks like that `rt_amp_spk_rtd_init()` will never be called with `rt7-something-`

bardliao replied:
We could have rt7xx + rt13xx combination case. For example, rt712 + rt1316 in a single dai link. In that case, rt1316 will call `rt_amp_spk_rtd_init` and `rt_amp_spk_rtd_init` will check all codec dais in the dai link which are rt712 and rt1316.
IOW, `rt_amp_spk_rtd_init` is called by rt13xx, but it will check all amps in the dai link which may contain rt7xx. And we need to skip the non-rt13xx amps in the dai link.

ujfalusi replied:
I don't get this (I got what you are saying, but)...
if we have multiple rt13xx in a single dai link then we add the entries multiple times?
The `rt_amp_spk_rtd_init` will be called for each dais, no?
If I look closer, the other `spk_rtd_init`s are adding a single entry unconditionally, they don't scan for all dais.

I'm sure this works, but it is a bit holistic for me ;)

bardliao replied:
> I don't get this (I got what you are saying, but)... if we have multiple rt13xx in a single dai link then we add the entries multiple times? The `rt_amp_spk_rtd_init` will be called for each dais, no? If I look closer, the other `spk_rtd_init`s are adding a single entry unconditionally, they don't scan for all dais.
> 

rt13xx will not be added multiple times, we check it with
		if (strstr(card->components, c_temp))
			continue;

rt7xx add a single entry unconditionally because the rt7xx_spk_rtd_init is specific for the codec, and we don't expect that there will be more than one rt7xx in a link.
However, rt_amp_spk_rtd_init is for all rt13xx amps, and it is possible we combine two different rt13xx amps like rt1316 + rt1308 together. It does not likely happen, but it is harmless to support this corner case.

And I am thinking maybe someday we can use a single spk_rtd_init for all kinds of amp. And we definitely need to check all codecs in the dai lnik.

> I'm sure this works, but it is a bit holistic for me ;)

ujfalusi replied:
@bardliao, right, I see. What is confusing is the fact that the `sof_sdw_dai_info.rtd_init()` callback has wrong name and scope (imho), it should be
```
int (*dai_init)(struct snd_soc_pcm_runtime *rtd, struct snd_soc_dai *dai);
bool dai_init_done; /* Indicate that the dai_init callback is done */
```
and deal with the given dai only.

bardliao replied:
> @bardliao, right, I see. What is confusing is the fact that the `sof_sdw_dai_info.rtd_init()` callback has wrong name and scope (imho), it should be
> 
> ```
> int (*dai_init)(struct snd_soc_pcm_runtime *rtd, struct snd_soc_dai *dai);
> bool dai_init_done; /* Indicate that the dai_init callback is done */
> ```
> 
> and deal with the given dai only.

Good suggestion, I will do it later.

ujfalusi reviewed (commented):
@bardliao, looks good, I think but I find the chage in rt_amp_spk_rtd_init() a bit confusing

ujfalusi reviewed (dismissed):
I still have a feeling that things are not too defined and we will need to add more duck-tape to hold it together.
Given the confidence from @bardliao, I can approve it as it brings some cleanup.

charleskeepax reviewed (approved):
Bits touching the Cirrus stuff all looks good to me.

plbossart reviewed (dismissed):
i share @ujfalusi's concern that we will have to revisit this, but at the same time we need to fix the tags to allow for the known RT712 configurations to be supported, so let's merge this and update later, if needed.

plbossart commented:
Argh, conflict! @bardliao can you rebase?

bardliao committed:
ASoC: Intel: sof_sdw: use .controls/.widgets to add controls/widgets

sof_sdw_rtd_init() will add the controls and widgets if we set them in
the codec_info. Move the additions from .rtd_init callback to
sof_sdw_rtd_init().

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao commented:
@plbossart @ujfalusi On the second thought, I would like to address @ujfalusi's comment in this PR.

bardliao committed:
ASoC: Intel: sof_sdw: add dai parameter to rtd_init callback

A generic .rtd_init() callback could be used by different dais. It is
useful to pass dai parameter to the callback. The dai parameter will be
used in the follow up commit. No functional change here.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: sof_sdw_rt_amp: use dai parameter

get_codec_name_and_route() get dai name from the first codec dai in a dai
link. However, the dai may not always be the first codec dai. Use the
dai paramter to make sure get_codec_name_and_route() refers to the
correct dai.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao commented:
@ujfalusi @plbossart rebased and updated.

charleskeepax commented on sound/soc/intel/boards/sof_sdw_rt_amp.c:
```diff
@@ -185,38 +173,24 @@ static const struct snd_soc_dapm_route *get_codec_name_and_route(struct snd_soc_
+	for_each_rtd_codec_dais(rtd, i, codec_dai) {
 		if (strstr(dai->component->name_prefix, "-1"))
 			ret = snd_soc_dapm_add_routes(&card->dapm, rt_amp_map, 2);
 		else if (strstr(dai->component->name_prefix, "-2"))
```
The changes to this loop look wrong. Previously those two strstr's checked against each DAI on the RTD now they repeatedly check against the passed in DAI.

bardliao replied:
> The changes to this loop look wrong. Previously those two strstr's checked against each DAI on the RTD now they repeatedly check against the passed in DAI.

Oh, yes. Thanks for point this out.

ujfalusi reviewed (approved):
@bardliao, I think there is a potential to simplify things further with something like this?
```diff
diff --git a/sound/soc/intel/boards/sof_sdw_rt711.c b/sound/soc/intel/boards/sof_sdw_rt711.c
index 7e54fc5cbe09..763be6d42bd6 100644
--- a/sound/soc/intel/boards/sof_sdw_rt711.c
+++ b/sound/soc/intel/boards/sof_sdw_rt711.c
@@ -69,24 +69,15 @@ static struct snd_soc_jack_pin rt711_jack_pins[] = {
 	},
 };
 
-static const char * const jack_codecs[] = {
-	"rt711"
-};
-
-int rt711_rtd_init(struct snd_soc_pcm_runtime *rtd)
+int rt711_rtd_init(struct snd_soc_pcm_runtime *rtd, struct snd_soc_dai *dai)
 {
 	struct snd_soc_card *card = rtd->card;
 	struct mc_private *ctx = snd_soc_card_get_drvdata(card);
-	struct snd_soc_dai *codec_dai;
 	struct snd_soc_component *component;
 	struct snd_soc_jack *jack;
 	int ret;
 
-	codec_dai = get_codec_dai_by_name(rtd, jack_codecs, ARRAY_SIZE(jack_codecs));
-	if (!codec_dai)
-		return -EINVAL;
-
-	component = codec_dai->component;
+	component = dai->component;
 	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
 					  "%s hs:rt711",
 					  card->components);
```
or perhaps this to the mix:
```c
	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
					  "%s hs:%s",
					  card->components, dai->name);
```
but I'm not sure if that is correct...

bardliao commented:
> @bardliao, I think there is a potential to simplify things further with something like this?
> 
> ```diff
> diff --git a/sound/soc/intel/boards/sof_sdw_rt711.c b/sound/soc/intel/boards/sof_sdw_rt711.c
> index 7e54fc5cbe09..763be6d42bd6 100644
> --- a/sound/soc/intel/boards/sof_sdw_rt711.c
> +++ b/sound/soc/intel/boards/sof_sdw_rt711.c
> @@ -69,24 +69,15 @@ static struct snd_soc_jack_pin rt711_jack_pins[] = {
>  	},
>  };
>  
> -static const char * const jack_codecs[] = {
> -	"rt711"
> -};
> -
> -int rt711_rtd_init(struct snd_soc_pcm_runtime *rtd)
> +int rt711_rtd_init(struct snd_soc_pcm_runtime *rtd, struct snd_soc_dai *dai)
>  {
>  	struct snd_soc_card *card = rtd->card;
>  	struct mc_private *ctx = snd_soc_card_get_drvdata(card);
> -	struct snd_soc_dai *codec_dai;
>  	struct snd_soc_component *component;
>  	struct snd_soc_jack *jack;
>  	int ret;
>  
> -	codec_dai = get_codec_dai_by_name(rtd, jack_codecs, ARRAY_SIZE(jack_codecs));
> -	if (!codec_dai)
> -		return -EINVAL;
> -
> -	component = codec_dai->component;
> +	component = dai->component;
>  	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
>  					  "%s hs:rt711",
>  					  card->components);
> ```
> 
> or perhaps this to the mix:
> 
> ```c
> 	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
> 					  "%s hs:%s",
> 					  card->components, dai->name);
> ```
> 
> but I'm not sure if that is correct...

Yes, we don't need get_codec_dai_by_name() anymore. I plan to remove get_codec_dai_by_name() in another PR.
